extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,685 @@
+          return fetchResponse.arrayBuffer().then(
+            bytes => {
+              if (promiseId != this.promiseId_) {
+                return Promise.reject(cancellation());
+              }
+              return this.validateAdResponse_(fetchResponse, bytes)
+                .then(valid => {
+                  if (promiseId != this.promiseId_) {
+                    return Promise.reject(cancellation());
+                  }
+                  return this.maybeRenderAmpAd_(valid, bytes);
+                });
+            });
+        } else {
+          return Promise.resolve(false);
+        }
+      })
+      .catch(error => this.promiseErrorHandler_(error));
+  }
+
+  /**
+   * Handles uncaught errors within promise flow.
+   * @param {string|Error} error
+   * @return {!Promise<string>}
+   * @private
+   */
+  promiseErrorHandler_(error) {
+    if (error && error.message && error.message.indexOf('amp-a4a: ') == 0) {
+      // caught previous call to promiseErrorHandler?  Infinite loop?
+      return Promise.reject(error);
+    }
+    if (error && error instanceof Error &&
+        error.message == cancellation().message) {
+      // Rethrow if cancellation
+      throw error;
+    }
+    // Returning promise reject should trigger unhandledrejection which will
+    // trigger reporting via src/error.js
+    const adQueryIdx = this.adUrl_ ? this.adUrl_.indexOf('?') : -1;
+    const state = {
+      'm': error ? error.message : '',
+      'tag': this.element.tagName,
+      'type': this.element.getAttribute('type'),
+      'au': adQueryIdx < 0 ? '' :
+            this.adUrl_.substring(adQueryIdx + 1, adQueryIdx + 251),
+    };
+    return Promise.reject(new Error('amp-a4a: ' + JSON.stringify(state)));
+  }
+
+  /** @override */
+  layoutCallback() {
+    // Promise may be null if element was determined to be invalid for A4A.
+    if (!this.adPromise_ || this.rendered_) {
+      return Promise.resolve();
+    }
+    // Layoutcallback only executes if ad is within viewport or render
+    // outside viewport returned true.  This is only relevant for non-AMP
+    // creatives which rendered via the buildCallback promise chain.  Ensure
+    // slot counts towards 3p loading count until we know that the creative is
+    // valid AMP.
+    this.timerId_ = incrementLoadingAds(this.getWin());
+    return this.adPromise_.then(rendered => {
+      if (!rendered) {
+        // Was not AMP creative so wrap in cross domain iframe.  layoutCallback
+        // has already executed so can do so immediately.
+        this.renderViaIframe_(true);
+      }
+      this.rendered_ = true;
+    }).catch(error => this.promiseErrorHandler_(error));
+  }
+
+  /** @override  */
+  unlayoutCallback() {
+    // Remove creative and reset to allow for creation of new ad.
+    if (!this.layoutMeasureExecuted_) {
+      return true;
+    }
+    // Iframe or shadow root attached as children.  Cannot delete shadowRoot
```

Comment:
It's in the W3C spec (https://www.w3.org/TR/2013/WD-components-intro-20130606/):  "Because you cannot remove a shadow root"...  however this is merely the proposal and a bit outdated but I have yet to find a way to delete it and it seems clear is the only option.